src/main/java/org/wildfly/security/auth/provider/JAASSecurityRealm.java

```diff
@@ -0,0 +1,244 @@
+        if (handler == null) {
+            // we will be using the default handler that only supports char[] and String credentials.
+            if (char[].class.isAssignableFrom(credentialType) || String.class.isAssignableFrom(credentialType) || ClearPassword.class.isAssignableFrom(credentialType)) {
+                return CredentialSupport.VERIFIABLE_ONLY;
+            }
+            else {
+                return CredentialSupport.UNSUPPORTED;
+            }
+        }
+        else {
+            // if a custom handler is set then the credential type is possibly verifiable.
+            return CredentialSupport.POSSIBLY_VERIFIABLE;
+        }
+    }
+
+    private class JAASRealmIdentity implements RealmIdentity {
+
+        private final Principal principal;
+
+        private Subject subject;
+
+        private JAASRealmIdentity(final Principal principal) {
+            this.principal = principal;
+        }
+
+        @Override
+        public Principal getPrincipal() throws RealmUnavailableException {
+            return this.principal;
+        }
+
+        @Override
+        public CredentialSupport getCredentialSupport(Class<?> credentialType) throws RealmUnavailableException {
+            if (handler == null) {
+                // we will be using the default handler that only supports char[] and String credentials.
+                if (char[].class.isAssignableFrom(credentialType) || String.class.isAssignableFrom(credentialType) || ClearPassword.class.isAssignableFrom(credentialType)) {
+                    return CredentialSupport.VERIFIABLE_ONLY;
+                }
+                else {
+                    return CredentialSupport.UNSUPPORTED;
+                }
+            }
+            else {
+                // if a custom handler is set then the credential type is possibly verifiable.
+                return CredentialSupport.POSSIBLY_VERIFIABLE;
+            }
+        }
+
+        @Override
+        public <C> C getCredential(Class<C> credentialType) throws RealmUnavailableException {
+            return null;
+        }
+
+        @Override
+        public boolean verifyCredential(Object credential) throws RealmUnavailableException {
+            boolean successfulLogin;
+            final CallbackHandler callbackHandler = this.createCallbackHandler(credential);
+            final Subject subject = new Subject();
+            final LoginContext context  = createLoginContext(loginConfiguration, subject, callbackHandler);
+
+            try {
+                context.login();
+                successfulLogin = true;
+                this.subject = subject;
+            } catch (LoginException le) {
+                ElytronMessages.log.debugJAASAuthenticationFailure(principal, le);
+                successfulLogin = false;
+            }
+            return successfulLogin;
+        }
+
+        @Override
+        public void dispose() {
```

Comment:
May need to look into that one a little more later but I would suggest the one in AuthenticatedRealmIdentity should call logout.